TypeCobol.Analysis/Graph/CfgAbstractInterpretation.Environment.cs

```diff
@@ -0,0 +1,341 @@
+                            _cyclicThreshold[b] = _cyclicExecutionThresold;
+                            if (!_cyclicRelation.TryGetValue(a, out var set))
+                            {
+                                set = new HashSet<int>();
+                                _cyclicRelation.Add(a, set);
+                            }
+                            set.Add(b);
+                        }
+                    }
+                    int na = Visited(a);
+                    if (na == d)
+                    {
+                        while (_stack.Count > 0 && _stack.Pop() != a)
+                        {
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Run a sub branch
+            /// </summary>
+            /// <param name="startBlock">The beginning branch</param>
+            /// <param name="stopBlock">The stopping branch which will not be executed.</param>
+            protected virtual void Run(BasicBlock<N, D> startBlock, BasicBlock<N, D> stopBlock)
+            {
+                _dfsStack.Push(startBlock);  
+                while (_dfsStack.Count > 0)
+                {
+                    BasicBlock<N, D> block = _dfsStack.Pop();
+                    if (block == stopBlock)
+                        return;
+                    if (CanExecute(block))
+                    {
+                        if (!_metricUpdated.Get(block.Index))
+                        {
+                            _metrics.NodeCount++;
+                            _metrics.EdgeCount += block.SuccessorEdges.Count;     
+                            if (block.Context != null) // This a control block.
+                                _metrics.ControlSubgraphCount++;
+                            _metricUpdated.Set(block.Index, true);
+                        }
+                        EnterBlock(block);
+                        IterateBlock(block);
+                        BasicBlock<N, D> nextFlowBlock = InterpretContext(block);                        
+                        LeaveBlock(block);
+                        if (nextFlowBlock == null)
+                        {
+                            foreach (var edge in block.SuccessorEdges)
+                            {
+                                var b = Cfg.SuccessorEdges[edge];
+                                if (CanAddExecution(block, b))
+                                {
+                                    _dfsStack.Push(b);
+                                }
+                            }
+                        }
+                        else
+                        {
+                            if (CanAddExecution(block, nextFlowBlock))
+                            {
+                                _dfsStack.Push(nextFlowBlock);
+                            }
+                        }
+                    }
+                }
+
+                bool CanExecute(BasicBlock<N, D> b)
```

Comment:
Why a `CanExecute` and `CanAddExecution` ? Why checking the value twice ?

Reply:
/// CanExecute is used during execution to check for a block that contributes to a cycle,
                /// if can be executed, that is to say if its cyclic threshold is not consumed.
                /// Other blocks are always executed.

                /// CanAddExecution Checks if a transition from one block to another can be pushed as to be executed.
                /// For a transition that can leads to a cycle, the cyclic threshold value is decreased, if the threshold
                /// is consumed the transition cannot be executed.